fpdf/encryption.py

```diff
@@ -258,30 +381,142 @@ def generate_owner_password(self):
             new_key = []
             for k in rc4key:
                 new_key.append(k ^ i)
-            result = ARC4().encrypt(new_key, result)
+            result = ARC4().encrypt(bytes(new_key), result)
         return bytes(result).hex()
 
-    def generate_user_password(self):
+    def generate_user_password(self) -> str:
         """
         PDF32000 reference - Algorithm 5: Computing the encryption dictionary's U (user password) value
         The security handler is only using revision 3 or 4, so the legacy r2 version is not implemented here
         """
         m = hashlib.new("md5", usedforsecurity=False)
         m.update(bytearray(self.DEFAULT_PADDING))
         m.update(bytes.fromhex(self.info_id))
-        result = m.digest()
+        result = bytearray(m.digest())
         key = self.k
         for i in range(20):
             new_key = []
             for k in key:
                 new_key.append(k ^ i)
-            result = ARC4().encrypt(new_key, result)
+            result = ARC4().encrypt(bytes(new_key), result)
         result.extend(
             (result[x] ^ self.DEFAULT_PADDING[x]) for x in range(16)
         )  # add 16 bytes of random padding
         return bytes(result).hex()
 
-    def generate_encryption_key(self):
+    @classmethod
+    def compute_hash(
+        cls: Type["StandardSecurityHandler"],
+        input_password: bytes,
+        salt: bytes,
+        user_key: bytes = bytearray(),
+    ) -> bytes:
+        """
+        Algorithm 2B - section 7.6.4.3.4 of the ISO 32000-2:2020
+        Applied on Security handlers revision 6
+        """
+        k = hashlib.sha256(input_password + salt + user_key).digest()
+        round_number = 0
+        while True:
+            round_number += 1
+            k1 = input_password + k + user_key
+            # Step (a + b)
+            cipher = Cipher(AES128(k[:16]), modes.CBC(k[16:32]))
+            encryptor = cipher.encryptor()
+            e = encryptor.update(k1 * 64) + encryptor.finalize()
+            # Step (c)
+            # remainder = int.from_bytes(e[:16], byteorder="big") % 3
+            remainder = sum(e[:16]) % 3
+            # Step (d)
+            if remainder == 0:
+                k = hashlib.sha256(e).digest()
+            elif remainder == 1:
+                k = hashlib.sha384(e).digest()
+            else:
+                k = hashlib.sha512(e).digest()
+            # Step (e)
+            if round_number >= 64 and e[-1] <= round_number - 32:
+                break
+
+        return k[:32]
+
+    def generate_user_password_rev6(self) -> None:
+        """
+        Generating the U (user password) and UE (user encryption)
+        for security handlers of revision 6
+        Algorithm 8 - Section 7.6.4.4.7 of the ISO 32000-2:2020
+        """
+        user_password = self.prepare_string(self.user_password)
+        if not user_password:
+            user_password = bytearray()
```

Comment:
Do you think this `if` would be worth a dedicated unit test?

Reply:
"test_encryption_aes256" already has a blank user password

Reply:
I know, I meant a unit test WITH a user password 🙂